File: src/action_server_template.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include <ros_template_programs/TimerAction.h>
#include <iostream>

namespace ros_template_programs {
    class ActionServer {
        private:
            ros::NodeHandle nh_;
            ros::NodeHandle pnh_;
            actionlib::SimpleActionServer<ros_template_programs::TimerAction> act_srv_;
            ros_template_programs::TimerGoal goal_;
            double init_time;

            void goalCb() ;
            void preemptCb();
            void controlCb();

        public:
            ActionServer( );
    };
}

void ros_template_programs::ActionServer::goalCb() {
    goal_ = *act_srv_.acceptNewGoal();
    ROS_INFO("[ %s ] Got a new target_time = %.2f", ros::this_node::getName().c_str(), goal_.target_time);
    controlCb();
    return;
}

void ros_template_programs::ActionServer::preemptCb() {
    ROS_INFO("[ %s ] Preempted\n", ros::this_node::getName().c_str() );
    ros_template_programs::TimerResult result;
    result.total_time = -1.0;
    act_srv_.setPreempted( result, "I got Preempted!" );
    return;
}

void ros_template_programs::ActionServer::controlCb() {
    if ( !act_srv_.isActive() || act_srv_.isPreemptRequested() ) return;
    ROS_INFO( "[ %s ]\nStart Timer", ros::this_node::getName().c_str() );

    ros_template_programs::TimerFeedback feedback;
    ros_template_programs::TimerResult result;

    double init_time = ros::Time::now().toSec();
    double curt_time = ros::Time::now().toSec();
    double target_time = goal_.target_time;
    feedback.elapsed_time = curt_time - init_time;

    ros::Rate loop_rate(2);
    while ( feedback.elapsed_time < target_time ) {
        curt_time = ros::Time::now().toSec();
        feedback.elapsed_time = curt_time - init_time;
        act_srv_.publishFeedback( feedback );
        ROS_INFO( "[ %s ]\npublishFeedback\nelapsed_time = %.2f\n", ros::this_node::getName().c_str(), feedback.elapsed_time );
        ros::spinOnce();
        loop_rate.sleep();
    }
    result.total_time = feedback.elapsed_time;
    ROS_INFO("[ %s ]\nIt's time to end\ntotal_time = %.2f\n", ros::this_node::getName().c_str(), result.total_time );
    act_srv_.setSucceeded( result );
    
    return;
}

ros_template_programs::ActionServer::ActionServer( ) : nh_(), pnh_("~"), act_srv_( nh_, "timer_action_server", false ) {
    act_srv_.registerGoalCallback( boost::bind( &ros_template_programs::ActionServer::goalCb, this ) );
    act_srv_.registerPreemptCallback ( boost::bind( &ros_template_programs::ActionServer::preemptCb, this ) );
    act_srv_.start();
    ROS_INFO( "[ %s ] Start the Server...\n", ros::this_node::getName().c_str() );
}

int main(int argc, char *argv[])  {
    ros::init(argc, argv, "action_server_template");
    ros_template_programs::ActionServer act_srv;
    ros::spin();
    return 0;
}

File: scripts/action_server_template.py
#! /usr/bin/env python
#coding:utf-8

import roslib; roslib.load_manifest('ros_template_programs')
import rospy
import actionlib
from ros_template_programs.msg import TimerAction
from ros_template_programs.msg import TimerGoal
from ros_template_programs.msg import TimerResult
from ros_template_programs.msg import TimerFeedback

class ActionServer():
    def __init__(self):
        self.act_srv = actionlib.SimpleActionServer( 'timer_action_server', TimerAction, execute_cb=None, auto_start=False )
        self.act_srv.register_goal_callback( self.goal_cb )
        self.act_srv.register_preempt_callback( self.preempt_cb )
        self.goal = TimerGoal()
        self.act_srv.start()
        rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\nStart the Server...\n' )

    def goal_cb(self):
        self.goal = self.act_srv.accept_new_goal();
        rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\n Got a new target_time = %.2f\n', self.goal.target_time )
        self.control_cb()
        return  

    def preempt_cb(self):
        rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\nPreempted...\n' )
        result = TimerResult()
        result.total_time = -1.0;
        self.act_srv.setPreempted( result, 'I got Preempted!' );
        return
    
    def control_cb(self):
        if (self.act_srv.is_active() == False or self.act_srv.is_preempt_requested() == True) : return
        rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\nStart Timer\n' )
        feedback = TimerFeedback()
        result = TimerResult()

        init_time = rospy.Time.now().to_sec()
        curt_time = init_time
        target_time = self.goal.target_time
        feedback.elapsed_time = curt_time - init_time

        loop_rate = rospy.Rate(2)
        while ( feedback.elapsed_time < target_time ) :
            curt_time = rospy.Time.now().to_sec()
            feedback.elapsed_time = curt_time - init_time;
            self.act_srv.publish_feedback( feedback )
            rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\npublishFeedback\nelapsed_time = %.2f\n', feedback.elapsed_time )
            loop_rate.sleep()
        result.total_time = feedback.elapsed_time
        rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\nIt is time to end\ntotal_time = %.2f\n', result.total_time )
        self.act_srv.set_succeeded( result )
        return

if __name__ == '__main__':
    try:
        rospy.init_node('action_server_template')
        node = ActionServer()
        rospy.spin()

    except rospy.ROSInterruptException: pass
